### ISSUE TITLE
request: simplified example of PKM

### ISSUE THREAD

**Issue by bionicles** [facebookresearch/XLM#177]
Hi,

Product Key Memory sounds super cool, and I want to make a port to TF 2.0 Keras. The problem is, the file is like 600+ lines long!

What would a simplified version look like? We don't need all the bells and whistles. Just a simple version would be super valuable for the research community to extend this work (and would probably yield more citations of the paper). 

I can't be the only one who wants to play with it but gets intimidated by the complexity of the example here...

**Comment by glample** [comment]
Hi,

It is true that many things in the code are not really useful, but more some previous hyper-parameters for different configurations that we tried and that did not necessarily work in our experiments. I'll be busy today & tomorrow, but I will try to make a branch with a minimalist version of the memory Sunday or Monday and get back to you.

In the meantime, I would say that you can forget almost anything in the file https://github.com/facebookresearch/XLM/blob/master/src/model/memory/query.py

The only thing here that really matters is `QueryMLP`. For that we tried a couple of things:
- have a separate network for each head
- have a same network for each head, and make the last layer `n_heads` larger to split it into `n_heads` queries
- use Grouped1DConvolution to try to make the computation of heads in parallel with CuDNN (this adds a bit of complexity in the code, but did not accelerate it at all. we clearly can remove this).

Overall, `QueryMLP` can be a simple 1-layer nn.Linear followed by an optional batchnorm.

This is the flat memory implementation, i.e. the baseline when you don't use product keys. This can be ignored and is just here for comparison: https://github.com/facebookresearch/XLM/blob/master/src/model/memory/memory.py#L409-L515

This is the initial implementation of PKM: https://github.com/facebookresearch/XLM/blob/master/src/model/memory/memory.py#L518-L632

But this is slower than `HashingMemoryProductFast`, so you can totally ignore `HashingMemoryProduct`.

Otherwise, a bunch of parameters are not really needed now since we identified what works well and does not work (or does not make any difference), such as: shuffle_indices, modulo_size, query_detach_input, score_subtract, query_net_learn (although these parameters do not make the code really bigger).

Hope this helps!

**Comment by glample** [comment]
Hi,

Here is a notebook with a minimalist implementation of the memory. It is the same configuration as the one used in the paper. https://github.com/facebookresearch/XLM/blob/master/PKM-layer.ipynb

**Comment by kobenaxie** [comment]
Hi @bionicles , Have you tried PKM in tensorflow (or keras) ? and any contributions for your experiments? In my tensorflow implementation, it seems make no difference ~~~

**Comment by glample** [comment]
@kobenaxie on which dataset are you trying the PKM? How large is it?

**Comment by kobenaxie** [comment]
@glample I used PKM in speech task not NLP, and the same configuration as the one used in your PKM paper.
